### PR TITLE
drivers: gpio: Add ALTERA PIO

### PR DESCRIPTION
**Pull request by gohshunjing**
Hi,

i have created a new PR, because i do not have write access to shilinte's repo.
This is the link to the stale PR.
https://github.com/zephyrproject-rtos/zephyr/pull/53766

I did a rebase and also added spinlocks for all write access to registers.
Do let me know if i have missed anything. Thank you.

Signed-off-by: Goh Shun Jing [shun.jing.goh@intel.com](mailto:shun.jing.goh@intel.com)

### CODE REVIEW

**Comment by gohshunjing** [comment]
I did a rebase to resolve conflicts in CODEOWNERS.
@nashif , @henrikbrixandersen any updates on this?
Please let me know, if i have missed anything. Thank you.

Best regards,
shun jing.

File: dts/bindings/gpio/altr,pio-1.0.yaml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    description: |
+      Direction can be set to Bidir, Input, InOut or Output.
+      It cannot be configure when during the runtime.
+      This is to make sure the driver is match with the hardware.
+    default: output
+    enum:
+      - "bidir"
+      - "input"
+      - "inout"
+      - "output"
```
Comment (henrikbrixandersen) [inline dts/bindings/gpio/altr,pio-1.0.yaml:25]: I take it by "hardware" you mean the synthesized FPGA IP core? If yes, I think it's worth mentioning (as hardware could also be the peripherals connected to the FPGA pads).

**Review by henrikbrixandersen — DISMISSED** [review]
A couple of minor comments, otherwise looks good:

File: drivers/gpio/gpio_altera_pio.c [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+
+static int gpio_altera_port_get_raw(const struct device *dev, uint32_t *value)
+{
+	const struct gpio_altera_config *cfg = dev->config;
+	uintptr_t reg_base = cfg->reg_base;
+	uint32_t addr;
+
+	addr = reg_base + ALTERA_AVALON_PIO_DATA_OFFSET;
+
+	if	(value == NULL)	{
```
[the remark this answers is not in the document]
Comment (gohshunjing) [reply drivers/gpio/gpio_altera_pio.c]: I have corrected the spacing. Thank you for the feedback.

File: dts/bindings/gpio/altr,pio-1.0.yaml (hunk 1 above)
Comment (gohshunjing) [reply dts/bindings/gpio/altr,pio-1.0.yaml]: i have changed the description as below:
  direction:
    type: string
    description: |
      Direction can be set to bidir, input, inout or output.
      It cannot be configure when during the runtime and should be
      set according to the synthesized FPGA design.

**Comment by gohshunjing** [comment]
rebased to resolve conflicts.

**Comment by gohshunjing** [comment]
Hi @henrikbrixandersen any updates for this PR?
Thank you for the support.
I have rebased again to resolve conflicts in drivers/gpio/CMakeLists.txt.

**Review by teburd — APPROVED** [review]
This looks fine to me

**Review by nashif — APPROVED** [review]

**Pull request merged**